src/inmanta/execute/scheduler.py

```diff
@@ -430,3 +470,220 @@ def run(self, compiler: "Compiler", statements: Sequence["Statement"], blocks: S
             raise RuntimeException(stmt.expression, "not all statements executed %s" % all_statements)
 
         return True
+
+
+@dataclass(frozen=True)
+class EntityRelationship:
+    """
+    Represent a relationship attribute of an entity.
+    """
+
+    fq_entity_name: str
+    relationship_name: str
+
+    @classmethod
+    def from_type_hint(cls, type_hint: "TypeHint") -> Tuple["EntityRelationship", "EntityRelationship"]:
+        first = cls(type_hint.first_type, type_hint.first_relation_name)
+        then = cls(type_hint.then_type, type_hint.then_relation_name)
+        return first, then
+
+    @classmethod
+    def from_delayed_result_variable(cls, drv: DelayedResultVariable[Any]) -> "EntityRelationship":
+        return cls(
+            fq_entity_name=drv.myself.get_type().get_full_name(),
+            relationship_name=drv.attribute.name,
+        )
+
+
+class PrioritisedDelayedResultVariableQueue:
+    """
+    A queue for DelayedResultVariables that is prioritized based on the
+    type hints passed to the Compiler. This queue will return elements
+    in the following order:
+
+    * First return the DelayedResultVariables, which are not an instance of
+      TempListVariable and that do not have an order constraint.
+    * Then return DelayedResultVariables with order constraint.
+      They are returned in an order that is valid with respect to the
+      constraints.
```

Comment:
I wonder if this could cause any issues. It is definitely the simplest and most intuitive approach, but it also means that simply adding a type hint that defines an order between two relations shifts both relations towards the end of the freezing order, which effectively orders them with respect to all other relations as well. I don't see any immediate issues with this myself but I think we need to carefully consider all consequences.
(@wouterdb)

Reply:
agreed, but I have no further insights either.